tensor_comprehensions/pybinds/tclib.cc

```diff
@@ -0,0 +1,586 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <stdint.h>
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include <glog/logging.h>
+
+#include <Python.h>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <torch/torch.h>
+
+#include "tc/aten/aten.h"
+#include "tc/aten/aten_autotuner.h"
+#include "tc/autotuner/genetic_search.h"
+#include "tc/autotuner/options_cache.h"
+#include "tc/core/cuda/cuda_backend.h"
+#include "tc/core/cuda/cuda_tc_executor.h"
+#include "tc/core/flags.h"
+#include "tc/core/tensor.h"
+#include "tc/lang/canonicalize.h"
+
+namespace tc {
+namespace python {
+
+namespace py = pybind11;
+
+namespace {
+void initGlog() {
+  static bool inited = false;
+  if (!inited) {
+    ::google::InitGoogleLogging("TC Python");
+    inited = true;
+  }
+}
+
+inline std::vector<tc::TensorInfo> getATenTensorsAsTensorInfo(
+    const py::tuple& pyTensors) {
+  std::vector<tc::TensorInfo> tensors;
+  for (auto& inp : pyTensors) {
+    tensors.push_back(tc::aten::toTensorInfo(inp.cast<at::Tensor>()));
+  }
+  return tensors;
+}
+
+inline std::vector<at::Tensor> getATenTensors(const py::tuple& pyTensors) {
+  std::vector<at::Tensor> atTensors;
+  for (auto& inp : pyTensors) {
+    atTensors.push_back(inp.cast<at::Tensor>());
+  }
+  return atTensors;
+}
+
+template <typename VoidPtr>
+inline std::vector<VoidPtr> getATenTensorsAsRawPtrs(
+    const py::tuple& pyTensors) {
+  std::vector<VoidPtr> res;
+  for (auto& inp : pyTensors) {
+    res.push_back(static_cast<VoidPtr>(inp.cast<at::Tensor>().data_ptr()));
+  }
+  return res;
+}
+
+inline py::list convertToPyObjects(const std::vector<at::Tensor>& tensors) {
+  py::list outputs;
+  for (auto& tensor : tensors) {
+    outputs.append(py::cast(torch::autograd::make_variable(tensor)));
+  }
+  return outputs;
+}
+} // namespace
+
+/**
+ * This struct serves the purpose of memoizing the compiled TcExecutors
+ * Since PyTorch's autograd is essentially stateless we cannot even store a
+ * pointer that corresponds to our invariants (TC, def, input sizes + strides).
+ * So we have to implement our own compilation cache.
+ *
+ * We want this to be lightweight so we cannot afford:
+ *   1. python-side dictionary manipulations,
+ *   2. parsing the TC string more than necessary (compilation and new
+ *      allocations are acceptable)
+ *   3. large string hashing
+ *
+ * Creating a key must be cheap, we cannot afford multiple conversions on the
+ * way from ATen tensors to TensorInfo.
+ */
+struct CompilationCache {
```

Comment:
I'm not a fa of providing extra functionality in the bindings only, but this is strictly better that what we had before